core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/messaginglayer/CongestionControl.scala

```diff
@@ -0,0 +1,64 @@
+package edu.uci.ics.amber.engine.architecture.messaginglayer
+
+import edu.uci.ics.amber.engine.architecture.messaginglayer.CongestionControl.{
+  maxWindowSize,
+  minWindowSize,
+  timeGap
+}
+import edu.uci.ics.amber.engine.architecture.messaginglayer.NetworkSenderActor.NetworkMessage
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+
+object CongestionControl {
+  final val maxWindowSize = 64
+  final val minWindowSize = 2
+  final val expendThreshold = 4
+  final val sendingTimeout: FiniteDuration = 30.seconds
+  final val timeGap = 3000 // 3s
+}
+
+class CongestionControl {
+
+  var ssThreshold = 16
+  var windowSize = minWindowSize
+  var sentTimeStamp = 0L
+  val toBeSent = new mutable.Queue[NetworkMessage]
+  val messageBuffer = new ArrayBuffer[NetworkMessage]()
+  private val inTransit = new mutable.LongMap[NetworkMessage]
+
+  def canBeSent(data: NetworkMessage): Boolean = {
+    if (inTransit.size < windowSize) {
+      sentTimeStamp = System.currentTimeMillis()
+      inTransit(data.messageID) = data
+      true
+    } else {
+      //println(s"queued $data")
+      toBeSent.enqueue(data)
+      false
+    }
+  }
+
+  def ack(id: Long): Array[NetworkMessage] = {
+    messageBuffer.clear()
```

Comment:
let's not clear the messageBuffer. If the earlier messageBuffer is still on its way, the receiving actor will get an empty array.

Reply:
In fact I think messageBuffer should be a local variable here. Create it using `new` before line 56.

Reply:
I defined it as a member because I think we should avoid object creation for each ack() call.